apis/s3/src/main/java/org/jclouds/s3/filters/RequestAuthorizeSignatureV4.java

```diff
@@ -0,0 +1,461 @@
+import java.net.URI;
+import java.net.URLEncoder;
+import java.security.DigestInputStream;
+import java.security.InvalidKeyException;
+import java.security.MessageDigest;
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static com.google.common.base.Charsets.UTF_8;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.io.BaseEncoding.base16;
+import static com.google.common.io.ByteStreams.readBytes;
+import static com.google.common.net.HttpHeaders.*;
+import static org.jclouds.aws.reference.AWSConstants.PROPERTY_HEADER_TAG;
+import static org.jclouds.crypto.Macs.asByteProcessor;
+import static org.jclouds.http.utils.Queries.queryParser;
+import static org.jclouds.s3.reference.S3Constants.PROPERTY_S3_VIRTUAL_HOST_BUCKETS;
+import static org.jclouds.util.Strings2.toInputStream;
+
+public class RequestAuthorizeSignatureV4 implements RequestAuthorizeSignature {
+    private static final String DEFAULT_ENCODING = "UTF-8";
+
+    /**
+     * Regex which matches any of the sequences that we need to fix up after
+     * URLEncoder.encode().
+     */
+    private static final Pattern ENCODED_CHARACTERS_PATTERN;
+
+    static {
+        StringBuilder pattern = new StringBuilder();
+
+        pattern
+            .append(Pattern.quote("+"))
+            .append("|")
+            .append(Pattern.quote("*"))
+            .append("|")
+            .append(Pattern.quote("%7E"))
+            .append("|")
+            .append(Pattern.quote("%2F"));
+
+        ENCODED_CHARACTERS_PATTERN = Pattern.compile(pattern.toString());
+    }
+
+    // Specifying a default for how to parse the service and region in this way allows
+    // tests or other downstream services to not have to use guice overrides.
+    @ImplementedBy(ServiceAndRegion.AWSServiceAndRegion.class)
+    public interface ServiceAndRegion {
+        String service();
+
+        String region(String host);
+
+        static final class AWSServiceAndRegion implements ServiceAndRegion {
+            private final String service;
+
+            @Inject
+            AWSServiceAndRegion(ProviderMetadata provider) {
+                this(provider.getEndpoint());
+            }
+
+            AWSServiceAndRegion(String endpoint) {
+                this.service = AwsHostNameUtils.parseServiceName(URI.create(checkNotNull(endpoint, "endpoint")));
+            }
+
+            @Override
+            public String service() {
+                return service;
+            }
+
+            @Override
+            public String region(String host) {
+                return AwsHostNameUtils.parseRegionName(host, service());
+            }
+        }
+    }
+
+    private final SignatureWire signatureWire;
+    private final String headerTag;
+    //    private final String apiVersion;
```

Comment:
copy from ```org.jclouds.aws.filters.FormSignerV4```
current, s3 api havn't any api version parameters, it's use in ec2 or other aws api...